libraries/AP_FETtecOneWire/AP_FETtecOneWire.h

```diff
@@ -0,0 +1,253 @@
+
+    /**
+        Resets a pending pull request
+    */
+    void pull_reset();
+
+    /**
+        Pulls a complete request between flight controller and ESC
+        @param esc_id id of the ESC
+        @param command 8bit array containing the command that should be send including the possible payload
+        @param response 8bit array where the response will be stored in
+        @param return_full_frame can be return_type::RESPONSE or return_type::FULL_FRAME
+        @param req_len transmit request length
+        @return true if the request is completed, false if dont
+    */
+    bool pull_command(const uint8_t esc_id, const uint8_t *command, uint8_t *response, return_type return_full_frame, const uint8_t req_len);
+
+    /**
+        Scans for all ESCs in bus. Configures fast-throttle and telemetry for the ones found.
+        Should be periodically called until _scan.state == scan_state_t::DONE
+    */
+    void scan_escs();
+
+    /**
+        configure the fast-throttle command.
+        Should be called once after scan_escs() is completted and before config_escs()
+    */
+    void config_fast_throttle();
+
+#if HAL_WITH_ESC_TELEM
+    /**
+        increment message packet count for every ESC
+    */
+    void inc_send_msg_count();
+
+    /**
+        calculates tx (outgoing packets) error-rate by converting the CRC error counts reported by the ESCs into percentage
+        @param esc_id id of ESC, that the error is calculated for
+        @param esc_error_count the error count given by the esc
+        @return the error in percent
+    */
+    float calc_tx_crc_error_perc(const uint8_t esc_id, uint16_t esc_error_count);
+
+    /**
+        if init is complete checks if the requested telemetry is available.
+        @param t telemetry datastructure where the read telemetry will be stored in.
+        @param centi_erpm 16bit centi-eRPM value returned from the ESC
+        @param tx_err_count Ardupilot->ESC communication CRC error counter
+        @param tlm_from_id receives the ID from the ESC that has respond with its telemetry
+        @return 1 if CRC is correct, 2 on CRC mismatch, 0 on waiting for answer
+    */
+    receive_response decode_single_esc_telemetry(TelemetryData& t, int16_t& centi_erpm, uint16_t& tx_err_count, uint8_t &tlm_from_id);
+#endif
+
+    /**
+        if init is complete sends a single fast-throttle frame containing the throttle for all found OneWire ESCs.
+        @param motor_values a 16bit array containing the throttle values that should be sent to the motors. 0-2000 where 1001-2000 is positive rotation and 0-999 reversed rotation
+        @param tlm_request the ESC to request telemetry from (-1 for no telemetry, 0 for ESC1, 1 for ESC2, 2 for ESC3, ...)
+    */
+    void escs_set_values(const uint16_t *motor_values, const int8_t tlm_request);
+
+    static constexpr uint8_t SERIAL_NR_BITWIDTH = 12;
+
+    typedef struct FETtecOneWireESC

```

Comment:
```suggestion
    class FETtecOneWireESC

```

and augment this with everything else that's of size MOTOR_COUNT_MAX.

Also add
```
    ESCState state;
```

where
```
enum class ESCState {
    UNKNOWN = 0,
    SENT_BOOT_COMMAND = 1,
    SENT_FW_VER_REQUEST = 2,
    SENT_SN_REQUEST = 3,
    .
    .
    RUNNING = 4,
};
```